src/Transport/Receiving/MessagePump.cs

```diff
@@ -203,24 +203,24 @@ async Task InnerProcessMessage(Task<Message> receiveTask)
             try
             {
                 using (var receiveCancellationTokenSource = new CancellationTokenSource())
+                using (var transaction = CreateTransaction())
                 {
-                    var transportTransaction = CreateTransportTransaction(message.PartitionKey);
+                    var transportTransaction = CreateTransportTransaction(message.PartitionKey, transaction);
 
                     var contextBag = new ContextBag();
                     contextBag.Set(message);
 
-                    var messageContext = new MessageContext(messageId, headers, body, transportTransaction, receiveCancellationTokenSource, contextBag);
+                    var messageContext = new MessageContext(messageId, headers, body, transportTransaction,
+                        receiveCancellationTokenSource, contextBag);
 
-                    using (var scope = CreateTransactionScope())
-                    {
-                        await onMessage(messageContext).ConfigureAwait(false);
+                    await onMessage(messageContext).ConfigureAwait(false);
 
-                        if (receiveCancellationTokenSource.IsCancellationRequested == false)
-                        {
-                            await receiver.SafeCompleteAsync(pushSettings.RequiredTransactionMode, lockToken).ConfigureAwait(false);
+                    if (receiveCancellationTokenSource.IsCancellationRequested == false)
+                    {
+                        await receiver.SafeCompleteAsync(pushSettings.RequiredTransactionMode, lockToken, transaction)
+                            .ConfigureAwait(false);
 
-                            scope?.Complete();
-                        }
+                        transaction?.Commit();
                     }
 
                     if (receiveCancellationTokenSource.IsCancellationRequested)
```

Comment:
Apparently, when the user requested a cancellation the behavior was always to make the message immediately reappear, therefore abandoning the current message has to happen outside the transaction. I wonder if an explicit Rollback would be good here or is it implicitly clear because the transaction is disposed without ever having called `Commit` (that is how the previous code worked too).

Reply:
I'd vote for an explicit Rollback personally